src/Common/FileSegment.cpp

```diff
@@ -0,0 +1,397 @@
+{
+    if (download_state == State::DOWNLOADED)
+        reserved_size = downloaded_size = size_;
+    else if (download_state != State::EMPTY)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Can create cell with either DOWNLOADED or EMPTY state");
+}
+
+FileSegment::State FileSegment::state() const
+{
+    std::lock_guard segment_lock(mutex);
+    return download_state;
+}
+
+size_t FileSegment::downloadOffset() const
+{
+    std::lock_guard segment_lock(mutex);
+    return range().left + downloaded_size;
+}
+
+String FileSegment::getCallerId()
+{
+    if (!CurrentThread::isInitialized() || CurrentThread::getQueryId().size == 0)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Cannot use cache without query id");
+
+    return CurrentThread::getQueryId().toString() + ":" + toString(getThreadId());
+}
+
+String FileSegment::getOrSetDownloader()
+{
+    std::lock_guard segment_lock(mutex);
+
+    if (downloader_id.empty())
+    {
+        if (download_state != State::EMPTY
+            && download_state != State::PARTIALLY_DOWNLOADED)
+            throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+                            "Can set downloader only for file segment with state EMPTY or PARTIALLY_DOWNLOADED, but got: {}",
+                            download_state);
+
+        downloader_id = getCallerId();
+        LOG_TEST(log, "Set downloader: {}, prev state: {}", downloader_id, stateToString(download_state));
+        download_state = State::DOWNLOADING;
+    }
+    else if (downloader_id == getCallerId())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Attempt to set the same downloader for segment {} for the second time", range().toString());
+
+    LOG_TEST(log, "Returning with downloader: {} and state: {}", downloader_id, stateToString(download_state));
+    return downloader_id;
+}
+
+String FileSegment::getDownloader() const
+{
+    std::lock_guard segment_lock(mutex);
+    return downloader_id;
+}
+
+bool FileSegment::isDownloader() const
+{
+    std::lock_guard segment_lock(mutex);
+    LOG_TEST(log, "Checking for current downloader. Caller: {}, downloader: {}, current state: {}", getCallerId(), downloader_id, stateToString(download_state));
+    return getCallerId() == downloader_id;
+}
+
+FileSegment::RemoteFileReaderPtr FileSegment::getRemoteFileReader()
+{
+    if (!isDownloader())
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Only downloader can use remote filesystem file reader");
+
+    return remote_file_reader;
+}
+
+void FileSegment::setRemoteFileReader(RemoteFileReaderPtr remote_file_reader_)
+{
+    if (!isDownloader())
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Only downloader can use remote filesystem file reader");
+
+    if (remote_file_reader)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Remote file reader already exists");
+
+    remote_file_reader = remote_file_reader_;
+}
+
+void FileSegment::write(const char * from, size_t size)
+{
```

Comment:
Shouldn't the lock be taken here?

Reply:
No, it should not. Just below there is a check `if (!isDownloader())` -- it checks if current thread is a downloader. Only one thread can be a downloader and a new downloader is not chosen until previous one is active. (And `write` method cannot be called concurrently anyway, this check for `isDownloader` is more like a logical error check).